crates/exex/exex/src/wal/storage.rs

```diff
@@ -0,0 +1,200 @@
+        };
+
+        Some(start..=end)
+    }
+
+    /// Removes notifications from the storage according to the given range.
+    pub(super) fn remove_notifications(
+        &mut self,
+        range: RemoveNotificationsRange,
+    ) -> eyre::Result<Vec<ExExNotification>> {
+        let adjusted_range = match range {
+            RemoveNotificationsRange::FromFileId(from_file_id) => {
+                self.adjust_file_range(from_file_id..)
+            }
+            RemoveNotificationsRange::ToFileId(to_file_id) => self.adjust_file_range(..to_file_id),
+        };
+        let Some(adjusted_range) = adjusted_range else { return Ok(Vec::new()) };
+
+        let removed_notifications =
+            self.iter_notifications(adjusted_range).collect::<eyre::Result<Vec<_>>>()?;
+
+        for (id, _) in &removed_notifications {
+            debug!(?id, "Removing notification from the storage");
+            reth_fs_util::remove_file(self.file_path(*id))?;
+        }
+
+        match range {
+            RemoveNotificationsRange::FromFileId(from_file_id) => {
+                self.max_id = from_file_id.checked_sub(1)
+            }
+            RemoveNotificationsRange::ToFileId(to_file_id) => self.min_id = Some(to_file_id),
+        };
+
+        Ok(removed_notifications.into_iter().map(|(_, notification)| notification).collect())
+    }
+
+    pub(super) fn iter_notifications(
+        &self,
+        range: impl RangeBounds<u64>,
+    ) -> Box<dyn Iterator<Item = eyre::Result<(u64, ExExNotification)>> + '_> {
+        let Some(range) = self.adjust_file_range(range) else {
+            return Box::new(std::iter::empty())
+        };
+
+        Box::new(
+            range.map(move |id| self.read_notification(id).map(|notification| (id, notification))),
+        )
+    }
+
+    /// Reads the notification from the file with the given id.
+    pub(super) fn read_notification(&self, file_id: u64) -> eyre::Result<ExExNotification> {
+        debug!(?file_id, "Reading notification from WAL");
+
+        let file_path = self.file_path(file_id);
+        let mut file = File::open(&file_path)?;
+        read_notification(&mut file)
+    }
+
+    /// Writes the notification to the file with the given id.
+    pub(super) fn write_notification(
+        &mut self,
+        notification: &ExExNotification,
+    ) -> eyre::Result<u64> {
+        let file_id = self.max_id.map_or(0, |id| id + 1);
+        self.min_id = self.min_id.map_or(Some(file_id), |min_id| Some(min_id.min(file_id)));
+        self.max_id = self.max_id.map_or(Some(file_id), |max_id| Some(max_id.max(file_id)));
+
+        debug!(?file_id, "Writing notification to WAL");
+
+        let file_path = self.file_path(file_id);
+        let mut file = File::create_new(&file_path)?;
+        write_notification(&mut file, notification)?;
+
+        Ok(file_id)
+    }
+}
+
+pub(super) enum RemoveNotificationsRange {
+    /// Remove notifications from the given file ID, inclusive.
+    FromFileId(u64),
```

Comment:
this comment is very confusing, does this remove individual notification from a file?